File: s3-inf-ppo/laboratoria/lab09/java/src/Zoo/Animals/Animal.java
package Zoo.Animals;

import Zoo.Log;

import java.util.Arrays;

abstract public class Animal {

    private String name;
    private boolean hungry = true;

    Animal(String name) {
        this.name = name;
    }

    abstract String[] getDiet();

    public void feed(String food) throws Exception {
        if(!Arrays.asList(getDiet()).contains(food)) {
            throw new Exception(getName() + " doesn't eat " + food + ".");
        }

        hungry = true;
        Log.info(getName() + " fed.");
    }

    public String getName() {
        return getSpecies() + " " + name;
    }

    private String getSpecies() {
        return getClass().getSimpleName();
    }

}

File: s3-inf-ppo/laboratoria/lab09/cs/lab09/Animals/Animal.cs
using System;
using System.Collections.Generic;

namespace lab09.Animals
{
    abstract class Animal
    {
        public string Name { get; set; }
        public bool Hungry { get; set; } = true;

        public Animal(string name)
        {
            Name = name;
        }

        public abstract List<string> GetDiet();

        public void Feed(string food)
        {
            if (!GetDiet().Contains(food))
            {
                throw new Exception(GetName() + " doesn't eat " + food + ".");
            }

            Hungry = true;
            Log.Info(GetName() + " fed.");
        }

        public string GetName()
        {
            return GetSpecies() + " " + Name;
        }

        private string GetSpecies()
        {
            return GetType().Name;
        }
    }
}
